rs/nervous_system/integration_tests/src/nervous_system_agent_helpers.rs

```diff
@@ -0,0 +1,538 @@
+
+                let proposal = match proposal_result {
+                    Ok(proposal) => proposal,
+                    Err(user_error) => {
+                        // Upgrading SNS Governance results in the proposal info temporarily not
+                        // being available due to the canister being stopped. This requires
+                        // more attempts to get the proposal info to find out if the proposal
+                        // actually got executed.
+                        if agent.is_canister_stopped_error(&user_error) {
+                            continue;
+                        } else {
+                            panic!("Error getting proposal: {:#?}", user_error);
+                        }
+                    }
+                };
+
+                let proposal = proposal
+                    .result
+                    .expect("GetProposalResponse.result must be set.");
+                let proposal_data = match proposal {
+                    get_proposal_response::Result::Error(err) => {
+                        panic!("Proposal data cannot be found: {:?}", err);
+                    }
+                    get_proposal_response::Result::Proposal(proposal_data) => proposal_data,
+                };
+                if proposal_data.executed_timestamp_seconds > 0 {
+                    return Ok(proposal_data);
+                }
+                proposal_data.failure_reason.clone().map_or(Ok(()), Err)?;
+                last_proposal_data = Some(proposal_data);
+            }
+            panic!(
+                "Looks like the SNS proposal {:?} is never going to be decided: {:#?}",
+                proposal_id, last_proposal_data
+            );
+        }
+
+        pub async fn get_proposal<C: CallCanisters>(
+            agent: &C,
+            governance_canister: GovernanceCanister,
+            proposal_id: ProposalId,
+        ) -> Result<GetProposalResponse, C::Error> {
+            governance_canister.get_proposal(agent, proposal_id).await
+        }
+
+        pub async fn get_caller_neuron<C: CallCanisters>(
+            agent: &C,
+            governance_canister: GovernanceCanister,
+        ) -> Result<NeuronId, C::Error> {
+            governance_canister
+                .list_neurons(
+                    agent,
+                    ListNeurons {
+                        limit: 1,
+                        of_principal: Some(agent.caller().unwrap().into()),
+                        start_page_at: None,
+                    },
+                )
+                .await
+                .map(|n| n.neurons[0].id.clone().expect("NeuronId must be set."))
+        }
+    }
+
+    pub mod swap {
+        use super::*;
+        use icrc_ledger_types::icrc1::{account::Account, transfer::TransferArg};
+
+        pub async fn refresh_buyer_tokens<C: CallCanisters>(
+            agent: &C,
+            sns_swap_canister: SwapCanister,
```

Comment:
I know the previous version of this code was written this way, but I'd suggest only using basic IC types for the function signatures, when possible. `SwapCanister` isn't a basic type, but `PrincipalId` and `CanisterId` are.